tool/planet/etcd.go

```diff
@@ -282,6 +287,91 @@ func etcdUpgrade(rollback bool) error {
 	return nil
 }
 
+// startWatchingEtcdMasters creates a control loop which polls etcd for the etcd cluster member list, and updates the
+// etcd gateway configuration with any changes. This keeps the etcd gateway load balancing in sync with the cluster.
+func startWatchingEtcdMasters(ctx context.Context) error {
+	config := etcdconf.Config{
+		Endpoints: []string{DefaultEtcdEndpoints},
+		KeyFile:   DefaultEtcdctlKeyFile,
+		CertFile:  DefaultEtcdctlCertFile,
+		CAFile:    DefaultEtcdctlCAFile,
+	}
+	cli, err := config.NewClientV3()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	go watchEtcdMasters(ctx, cli)
+	return nil
+}
+
+func watchEtcdMasters(ctx context.Context, client *etcdv3.Client) {
+	ticker := time.NewTicker(15 * time.Second)
+	defer ticker.Stop()
+
+	gateway := etcdGateway{}
+
+	for {
+		select {
+		case <-ticker.C:
+			err := gateway.resyncEtcdMasters(ctx, client)
+			if err != nil {
+				log.WithError(err).Warn("Error resyncing etcd master list.")
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+type etcdGateway struct {
+	clientURLs []string
```

Comment:
You don't really need a struct to keep track of client URLs - it could be as simple as a local variable in `watchEtcdMasters`:
```go
func watchEtcdMasters(ctx context.Context, client *etcdv3.Client) {
	ticker := time.NewTicker(15 * time.Second)
	defer ticker.Stop()

	var err error
	var clientURLs []string
	for {
		select {
		case <-ticker.C:
			clientURLs, err = resyncEtcdMasters(ctx, client, clientURLs)
			if err != nil {
				log.WithError(err).Warn("Error resyncing etcd master list.")
			}
		case <-ctx.Done():
			return
		}
	}
}

func resyncEtcdMasters(ctx context.Context, client *etcdv3.Client, clientURLs []string) (newClientURLs []string, err error) {
	memberList, err := client.MemberList(ctx)
	if err != nil {
		return clientURLs, trace.Wrap(err, "error retrieving member list")
	}

	newClientURLs, err := collectClientURLs(memberList)
	if err != nil {
		return clientURLs, trace.Wrap(err)
	}

	// Only rewrite the configuration if there are changes
	sort.Strings(newClientURLs)
	if reflect.DeepEqual(newClientURLs, clientURLs) {
		return clientURLs, nil
	}

	err = updateEtcdClientURLs(newClientURLs)
	if err != nil {
		return clientURLs, trace.Wrap(err)
	}

	return newClientURLs, nil
}
```

Reply:
I'm not sure I see how this makes it any simpler?

Reply:
I don't have a strong opinion on using a struct vs a local var here but I do like Dima's suggestion on encapsulating some logic into `collectClientURLs(memberList)` and `updateEtcdClientURLs(newClientURLs)` functions.